src/compiler/transformers/classFields.ts

```diff
@@ -506,6 +786,165 @@ namespace ts {
+            const text = getTextOfPropertyName(name) as string;
+            const weakMapName = createOptimisticUniqueName("_" + text.substring(1));
+            weakMapName.autoGenerateFlags |= GeneratedIdentifierFlags.ReservedInNestedScopes;
+            hoistVariableDeclaration(weakMapName);
+            env.set(name.escapedText, { placement: PrivateIdentifierPlacement.InstanceField, weakMapName });
+            (pendingExpressions || (pendingExpressions = [])).push(
+                createAssignment(
+                    weakMapName,
+                    createNew(
+                        createIdentifier("WeakMap"),
+                        /*typeArguments*/ undefined,
+                        []
+                    )
+                )
+            );
+        }
+
+        function accessPrivateIdentifier(name: PrivateIdentifier) {
+            for (let i = privateIdentifierEnvironmentStack.length - 1; i >= 0; --i) {
+                const env = privateIdentifierEnvironmentStack[i];
+                if (env.has(name.escapedText)) {
+                    return env.get(name.escapedText);
+                }
+            }
+            return undefined;
+        }
+
+
+        function wrapPrivateIdentifierForDestructuringTarget(node: PrivateIdentifierPropertyAccessExpression) {
+            return createPropertyAccess(
+                createObjectLiteral([
+                    createSetAccessor(
+                        /*decorators*/ undefined,
+                        /*modifiers*/ undefined,
+                        "value",
+                        [createParameter(
+                            /*decorators*/ undefined,
+                            /*modifiers*/ undefined,
+                            /*dotDotDotToken*/ undefined, "x",
+                            /*questionToken*/ undefined,
+                            /*type*/ undefined,
+                            /*initializer*/ undefined
+                        )],
+                        createBlock(
+                            [createExpressionStatement(
+                                visitNode(
+                                    createAssignment(node, createIdentifier("x")),
+                                    visitor
+                                )
+                            )]
+                        )
+                    )
+                ]),
+                "value"
+            );
+        }
+
+        function transformDestructuringAssignmentTarget(node: ArrayLiteralExpression | ObjectLiteralExpression) {
+            const hasPrivateIdentifiers = isArrayLiteralExpression(node) ?
+                forEach(node.elements, isPrivateIdentifierPropertyAccessExpression) :
+                forEach(node.properties, property => isPropertyAssignment(property) && isPrivateIdentifierPropertyAccessExpression(property.initializer));
+            if (!hasPrivateIdentifiers) {
+                return node;

```

Comment:
Don't forget to visit `node`, as it may have an element or a property with an initializer that itself contains public fields or private names. You should also add a test case for this to verify behavior and avoid regressions:

```ts
[a, [x.#b], c] = [1, [2], 3];
({ a, b: [x.#b], c } = { a: 1, b: [2], c: 3 });
```

If you do this here, you can exit `visitBinaryExpression` unconditionally rather than comparing the result.